File: linked_list/24.swap-nodes-in-pairs.cpp
/*
 * @lc app=leetcode id=24 lang=cpp
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* swapPairs(ListNode* head) {
    ListNode dummy, *curr = &dummy;
    curr->next = head;

    while (curr->next and curr->next->next) {
      ListNode *a = curr->next, *b = curr->next->next, *n = b->next;
      curr->next = b;
      b->next = a;
      a->next = n;
      curr = a;
    }

    return dummy.next;
  }
};
// @lc code=end

File: 24.swap-nodes-in-pairs.py
#
# @lc app=leetcode id=24 lang=python3
#
# [24] Swap Nodes in Pairs
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(-1, head)
        prev = dummy
        curr = head

        while curr and curr.next:
            first_node = curr
            second_node = curr.next
            next_node = curr.next.next

            prev.next = second_node
            second_node.next = first_node
            first_node.next = next_node

            prev = first_node
            curr = next_node

        return dummy.next
        
# @lc code=end

